src/network/networkd-sysctl.c

```diff
@@ -4,18 +4,209 @@
 #include <linux/if.h>
 #include <linux/if_arp.h>
 
+#include "sd-messages.h"
+
 #include "af-list.h"
+#include "cgroup-util.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-lldp-tx.h"
 #include "networkd-manager.h"
 #include "networkd-ndisc.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        sym_ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int sysctl_event_handler(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        /* Returning a negative value interrupts the ring buffer polling,
+         * so do it only in case of a fatal error like a version mismatch. */
+        if (we->version != 1)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                "Unexpected sysctl event, disabling sysctl monitoring: %d", we->version);
+
+        if (we->errorcode < 0) {
```

Comment:
better make this != 0?

in systemd we usually treat negative and positive errors the same. maybe just do the same here for robustness.

Reply:
log_warning_errno() accepts both anyway.